rclpy/test/test_action_client.py

```diff
@@ -0,0 +1,304 @@
+            future = ac.send_goal_async(
+                Fibonacci.Goal(),
+                feedback_callback=self.feedback_callback,
+                goal_uuid=goal_uuid)
+            rclpy.spin_until_future_complete(self.node, future, self.executor)
+
+            # Check the feedback was not received
+            self.assertEqual(self.feedback, None)
+        finally:
+            ac.destroy()
+
+    def test_send_goal_async_with_feedback_for_another_goal(self):
+        ac = ActionClient(self.node, Fibonacci, 'fibonacci')
+        try:
+            self.assertTrue(ac.wait_for_server(timeout_sec=2.0))
+
+            # Send a goal and then publish feedback
+            first_goal_uuid = UUID(uuid=list(uuid.uuid4().bytes))
+            future = ac.send_goal_async(
+                Fibonacci.Goal(),
+                feedback_callback=self.feedback_callback,
+                goal_uuid=first_goal_uuid)
+            rclpy.spin_until_future_complete(self.node, future, self.executor)
+
+            # Send another goal, but without a feedback callback
+            second_goal_uuid = UUID(uuid=list(uuid.uuid4().bytes))
+            future = ac.send_goal_async(
+                Fibonacci.Goal(),
+                goal_uuid=second_goal_uuid)
+            rclpy.spin_until_future_complete(self.node, future, self.executor)
+
+            # Publish feedback for the second goal
+            self.mock_action_server.publish_feedback(second_goal_uuid)
+            self.timed_spin(1.0)
+            self.assertEqual(self.feedback, None)
+            # Publish feedback for the first goal (with callback)
+            self.mock_action_server.publish_feedback(first_goal_uuid)
+            self.timed_spin(1.0)
+            self.assertNotEqual(self.feedback, None)
+        finally:
+            ac.destroy()
+
+    def test_send_goal_async_with_feedback_for_not_a_goal(self):
+        ac = ActionClient(self.node, Fibonacci, 'fibonacci')
+        try:
+            self.assertTrue(ac.wait_for_server(timeout_sec=2.0))
+
+            # Send a goal and then publish feedback
+            goal_uuid = UUID(uuid=list(uuid.uuid4().bytes))
+            future = ac.send_goal_async(
+                Fibonacci.Goal(),
+                feedback_callback=self.feedback_callback,
+                goal_uuid=goal_uuid)
+            rclpy.spin_until_future_complete(self.node, future, self.executor)
+
+            # Publish feedback for a non-existent goal ID
+            self.mock_action_server.publish_feedback(UUID(uuid=list(uuid.uuid4().bytes)))
+            self.timed_spin(1.0)
+            self.assertEqual(self.feedback, None)
+        finally:
+            ac.destroy()
+
+    def test_send_goal_async_no_server(self):
+        ac = ActionClient(self.node, Fibonacci, 'not_fibonacci')
+        try:
+            future = ac.send_goal_async(Fibonacci.Goal())
+            self.assertFalse(rclpy.spin_once(self.node, executor=self.executor, timeout_sec=2.0))
```

Comment:
`rclpy.spin_once()` always returns `None`, so the `assertFalse()` is unnecessary.

https://github.com/ros2/rclpy/blob/91996411deba25b662aaef08440a3103ed38c60c/rclpy/rclpy/__init__.py#L92-L93

Reply:
Also since this is being run with `pytest,` plain `assert` statements can be used instead of `unittest`'s `assertFalse()`. `pytest` does a pretty good job printing values from `assert` https://docs.pytest.org/en/latest/example/reportingdemo.html#tbreportdemo

No change required, it's up to you what style you prefer.

Reply:
Since there could be multiple things `spin_once` could spin on, use `timed_spin()` instead?

Reply:
Updated to use `timed_spin()` 37c98d6

---

I went with `unittest`'s style for consistency with the other tests I looked at in the package, but in retrospect there are a few that use plain `assert` as well. IMO I prefer the aesthetic of plain `assert`. What do you think about an overhaul of all the tests so they're consistent in a follow-up PR?

Reply:
Converting the tests from unittest to pytest style in a follow up PR sounds good to me